### ISSUE TITLE
regression: materialized view is missing rows, "since" is in the past, "upper" is empty

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#15185]
### What version of Materialize are you using?

 v0.27.0-alpha.23.post (469bd67f9)

### How did you install Materialize?

Docker image

### What is the issue?

The DeltaJoin scenario in the Platform Checks test has started failing reproducibly as follows:

```
ows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 49s 74s 63s
--
  | <stdin>:2:1: error: non-matching rows: expected:
  | [["1", "1", "1", "1"], ["2", "2", "2", "2"], ["3", "3", "3", "3"]]
  | got:
  | [["2", "2", "2", "2"], ["3", "3", "3", "3"]]
```

This means that a materialized view is missing a row that it should have after storaged was restarted: 

```
materialize=> select * from delta_join_view1;
 f1 | f2 | f3 | f4 
----+----+----+----
  2 |  2 |  2 |  2
  3 |  3 |  3 |  3
(2 rows)
```
Row `[1,1,1,1]` is missing. The row is present in the base tables and if one recreates the view from scratch the row is there:

```
materialize=> select * from "delta_join_view1A";
 f1 | f2 | f3 | f4 
----+----+----+----
  1 |  1 |  1 |  1
  2 |  2 |  2 |  2
  3 |  3 |  3 |  3
(3 rows)
```

The row the is missing is the one that was inserted in the table first. There were no deletions to the base tables. `EXPLAIN TIMESTAMP` shows something is off:
```
materialize=> explain timestamp for select * from "delta_join_view1";
                              Timestamp                              
---------------------------------------------------------------------
           query timestamp: 1664968962125 (2022-10-05 11:22:42.125) +
                     since:[1664967854000 (2022-10-05 11:04:14.000)]+
                     upper:[]                                       +
          global timestamp: 1664968962125 (2022-10-05 11:22:42.125) +
   can respond immediately: true                                    +
                                                                    +
 source materialize.public.delta_join_view1 (u5, storage):          +
             read frontier:[1664967854000 (2022-10-05 11:04:14.000)]+
            write frontier:[]                                       +
```

The `since` is stuck in the past and `upper` is empty.

To reproduce, start a `ci` machine using `bin/scratch` and then run:
```
while bash -c -e "./mzcompose down -v ; ./mzcompose run default --scenario=RestartEnvironmentdStoraged --check=DeltaJoin" ;  do : ; done
```

and give it a couple of minutes.

**Comment by philip-stoev** [comment]
This continues to happen as described.

**Comment by aljoscha** [comment]
I believe I diagnosed the missing data as well! I have a branch with more logging statements here: https://github.com/aljoscha/materialize/pull/new/storage-tease-out-mv-issues-take2. I also simplified the scenario a bit, but it exhibits the same faulty behavior. You can checkout my branch and repro using Philips instructions from above.

I'm appending a heavily annotated version of the log output that describes what's going on below [1] but it goes roughly like this:

1. we create two tables (`table1` and `table2`), along with an index for each table
2. we create a materialized view that joins `table1` and `table2`, the intention is that this is a delta join, and it seems to be using the two indexes as input rather than going all the way to a `persist_source` for the two tables
3. we restart `environmentd` and `storaged` a couple of times in between, but NOT `computed`
   - we can see in the logs how `environmentd` reconnects and how the `persist_source` is sometimes re-created but sometimes not during reconciliation
4.  on the final restart, the `persist_sink` (which is backing the materialized view) does ONLY get `2 2 2 2` on it's input, but not `1 1 1 1`
5. the `persist_sink` eventually gets `3 3 3 3` as input from the join
6. the `persist_sink` goes on to write out `3 3 3 3` AND (correctly) a retraction for `1 1 1 1`, because it is not in the "desired input"

**The last step is the correct behaviour for `persist_sink`, but it is not correct for the overall pipeline. We _do_ want to have that `1 1 1 1` in the output.**

It seems both `persist_source` and `persist_sink` are working as expected but somewhere in the "middle" we lose that `1 1 1 1` update in the process of restarting/reconciling.

[1]
```
2022-10-20T11:57:22.695190Z  INFO computed: about to bind to 0.0.0.0:2100
2022-10-20T11:57:22.695251Z  INFO computed: serving internal HTTP server on 0.0.0.0:6878
2022-10-20T11:57:22.696434Z  INFO mz_service::grpc: Starting to listen on 0.0.0.0:2100
2022-10-20T11:57:22.880595Z  INFO mz_service::grpc: GrpcServer: remote client connected
2022-10-20T11:57:22.881762Z  INFO mz_compute::communication: initialization complete worker=0

--aljoscha: this is the source for table1, that is fed into its index
2022-10-20T11:57:23.103085Z  INFO mz_storage::source::persist_source: persist_source u1/s39f8c082-5e3a-4dda-a399-75324c57c18a, initial as_of: Some(Antichain { elements: [1666267043042] })
2022-10-20T11:57:23.150709Z  INFO mz_storage::source::persist_source: persist_source u1, emitting update: (Ok(Row{[Int32(1), Int32(1)]}), 1666267043106, 1)

--aljoscha: this is the source for table2, that is fed into its index
2022-10-20T11:57:23.210984Z  INFO mz_storage::source::persist_source: persist_source u3/s62778c11-2287-458c-b9b6-1c65c9721971, initial as_of: Some(Antichain { elements: [1666267043132] })
2022-10-20T11:57:23.231965Z  INFO mz_storage::source::persist_source: persist_source u3, emitting update: (Ok(Row{[Int32(1), Int32(1)]}), 1666267043214, 1)

2022-10-20T11:57:26.183472Z  INFO mz_service::grpc: GrpcServer: remote client connected
2022-10-20T11:57:39.667091Z  INFO mz_service::grpc: GrpcServer: remote client connected

2022-10-20T11:57:43.518177Z  INFO mz_storage::source::persist_source: persist_source u3, emitting update: (Ok(Row{[Int32(2), Int32(2)]}), 1666267056574, 1)
2022-10-20T11:57:43.761151Z  INFO mz_storage::source::persist_source: persist_source u1, emitting update: (Ok(Row{[Int32(2), Int32(2)]}), 1666267056555, 1)

--aljoscha: this is the persist source that the sink uses to learn of successful writes
2022-10-20T11:57:51.600173Z  INFO mz_storage::source::persist_source: persist_source u5/s82a384a1-c68c-4383-b7a2-8ce47375c516, initial as_of: None

--aljoscha: our persist_sink for the materialized view
2022-10-20T11:57:51.600298Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: initial as_of: Antichain { elements: [1666267070000] }

--aljoscha: first bunch of "desired updates", the input that the sink must write out
2022-10-20T11:57:51.609980Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: desired updates: [(Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267070000, 1), (Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267070000, 1)], in-flight-batches: {}, desired_frontier: Antichain { elements: [0] }, batch_descriptions_frontier: Antichain { elements: [0] }, persist_frontier: Antichain { elements: [1666267070000] }

--aljoscha: we fail to write on the first try, because the upper is not as expected (side note: we could fix this if we want, but it's still correct)
2022-10-20T11:57:51.617299Z  INFO mz_compute::sink::persist_sink: persist_sink u5 writing ((SourceData(Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]})), ()), 1666267070000, 1), batch: Antichain { elements: [0] } -> Antichain { elements: [1666267071555] }
2022-10-20T11:57:51.617323Z  INFO mz_compute::sink::persist_sink: persist_sink u5 writing ((SourceData(Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]})), ()), 1666267070000, 1), batch: Antichain { elements: [0] } -> Antichain { elements: [1666267071555] }
2022-10-20T11:57:51.624029Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: append result for batch (Antichain { elements: [0] } -> Antichain { elements: [1666267071555] }): Err(Upper(Antichain { elements: [1666267070000] }))

--aljoscha: second attempt at writing succeeds
2022-10-20T11:57:51.624167Z  INFO mz_compute::sink::persist_sink: persist_sink u5 writing ((SourceData(Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]})), ()), 1666267070000, 1), batch: Antichain { elements: [1666267070000] } -> Antichain { elements: [1666267071555] }
2022-10-20T11:57:51.624189Z  INFO mz_compute::sink::persist_sink: persist_sink u5 writing ((SourceData(Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]})), ()), 1666267070000, 1), batch: Antichain { elements: [1666267070000] } -> Antichain { elements: [1666267071555] }
2022-10-20T11:57:51.630411Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: append result for batch (Antichain { elements: [1666267070000] } -> Antichain { elements: [1666267071555] }): Ok(())

--aljoscha: persist_sink learns about the updates it has written out
2022-10-20T11:57:51.643649Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: persist updates: [(Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267070000, 1), (Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267070000, 1)], in-flight-batches: {}, desired_frontier: Antichain { elements: [1666267071555] }, batch_descriptions_frontier: Antichain { elements: [1666267070001] }, persist_frontier: Antichain { elements: [1666267070000] }

--aljoscha: this must be the reconnected environmentd (the test scenario restarts environmentd, but not computed)
2022-10-20T11:57:54.678992Z  INFO mz_service::grpc: GrpcServer: remote client connected

--aljoscha: update for table1
2022-10-20T11:57:54.681040Z  INFO mz_storage::source::persist_source: persist_source u1, emitting update: (Ok(Row{[Int32(3), Int32(3)]}), 1666267071555, 1)

--aljoscha: persist_sink, and the persist_source that it uses are re-created
2022-10-20T11:57:55.043206Z  INFO mz_storage::source::persist_source: persist_source u5/s82a384a1-c68c-4383-b7a2-8ce47375c516, initial as_of: None
--aljoscha: note here, that the initial as_of is less than what it was before, this should not lead to incorrect output, because the sink is self-correcting, though!
2022-10-20T11:57:55.043318Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: initial as_of: Antichain { elements: [0] }

--aljoscha: this is the first incarnation of persist_sink noticing that it's being cancelled
2022-10-20T11:57:55.044778Z  INFO mz_compute::sink::persist_sink: persist_sink u5 exiting because token was taken

--aljoscha: the new incarnation of persist_sink started a new persist_source that emits what's in the shard
2022-10-20T11:57:55.067308Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267070000, 1)
2022-10-20T11:57:55.067339Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267070000, 1)

--aljoscha: the sink gets the desired updates again. NOTICE: the row for "1 1 1 1" is missing
2022-10-20T11:57:55.077457Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: desired updates: [(Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267056574, 1)], in-flight-batches: {}, desired_frontier: Antichain { elements: [0] }, batch_descriptions_frontier: Antichain { elements: [0] }, persist_frontier: Antichain { elements: [1666267070000] }

--aljoscha: the new incarnation of persist_sink "learns" about what was already written from it's persist_source
2022-10-20T11:57:55.077495Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: persist updates: [(Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267070000, 1), (Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267070000, 1)], in-flight-batches: {}, desired_frontier: Antichain { elements: [0] }, batch_descriptions_frontier: Antichain { elements: [0] }, persist_frontier: Antichain { elements: [1666267070000] }

--aljoscha: I _think_ this is the persist_source of a PEEK, we'll see several of these below as the test retries reading
2022-10-20T11:57:55.466290Z  INFO mz_storage::source::persist_source: persist_source u5/s82a384a1-c68c-4383-b7a2-8ce47375c516, initial as_of: Some(Antichain { elements: [1666267086554] })

--aljoscha: sink is not yet ready, though...
2022-10-20T11:57:55.535848Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086554] } not yet available for upper Antichain { elements: [1666267071555] } retrying in 64ms
2022-10-20T11:57:55.603835Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086554] } not yet available for upper Antichain { elements: [1666267071555] } retrying in 128ms
2022-10-20T11:57:55.725438Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086554] } not yet available for upper Antichain { elements: [1666267071555] } retrying in 256ms
2022-10-20T11:57:55.997563Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086554] } not yet available for upper Antichain { elements: [1666267071555] } retrying in 512ms
2022-10-20T11:57:56.527005Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086554] } not yet available for upper Antichain { elements: [1666267071555] } retrying in 1.024s
2022-10-20T11:57:57.456276Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086554] } not yet available for upper Antichain { elements: [1666267071555] } retrying in 2.048s

--aljoscha: update for table2
2022-10-20T11:57:59.453852Z  INFO mz_storage::source::persist_source: persist_source u3, emitting update: (Ok(Row{[Int32(3), Int32(3)]}), 1666267071603, 1)

--aljoscha: persist_sink learns of the new "3 3 3 3" row
2022-10-20T11:57:59.454319Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: desired updates: [(Ok(Row{[Int32(3), Int32(3), Int32(3), Int32(3)]}), 1666267071603, 1)], in-flight-batches: {}, desired_frontier: Antichain { elements: [0] }, batch_descriptions_frontier: Antichain { elements: [0] }, persist_frontier: Antichain { elements: [1666267071555] }

--aljoscha: can ignore, it's again trying an append that won't work
2022-10-20T11:57:59.454926Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: append result for batch (Antichain { elements: [0] } -> Antichain { elements: [1666267071555] }): Err(Upper(Antichain { elements: [1666267071555] }))

--aljoscha: persist_sink correctly writes out the new "3 3 3 3" AND a retraction for "1 1 1 1" because it's not in desired, that is we didn't get it as input
2022-10-20T11:57:59.455098Z  INFO mz_compute::sink::persist_sink: persist_sink u5 writing ((SourceData(Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]})), ()), 1666267071555, -1), batch: Antichain { elements: [1666267071555] } -> Antichain { elements: [1666267086555] }
2022-10-20T11:57:59.455116Z  INFO mz_compute::sink::persist_sink: persist_sink u5 writing ((SourceData(Ok(Row{[Int32(3), Int32(3), Int32(3), Int32(3)]})), ()), 1666267071603, 1), batch: Antichain { elements: [1666267071555] } -> Antichain { elements: [1666267086555] }
2022-10-20T11:57:59.460730Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: append result for batch (Antichain { elements: [1666267071555] } -> Antichain { elements: [1666267086555] }): Ok(())

--aljoscha: from now on, it's just PEEK results
--aljoscha: NOTE: we never get "1 1 1 1" as a desired input for the persist_sink
2022-10-20T11:57:59.687179Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086554, 1)
2022-10-20T11:57:59.687210Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267086554, 1)
2022-10-20T11:57:59.687463Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086554, -1)
2022-10-20T11:57:59.687476Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(3), Int32(3), Int32(3), Int32(3)]}), 1666267086554, 1)
2022-10-20T11:57:59.690129Z  INFO mz_storage::source::persist_source: persist_source u5/s82a384a1-c68c-4383-b7a2-8ce47375c516, initial as_of: Some(Antichain { elements: [1666267086554] })
2022-10-20T11:57:59.701034Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086554, 1)
2022-10-20T11:57:59.701056Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267086554, 1)
2022-10-20T11:57:59.701307Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086554, -1)
2022-10-20T11:57:59.701320Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(3), Int32(3), Int32(3), Int32(3)]}), 1666267086554, 1)
2022-10-20T11:57:59.766234Z  INFO mz_storage::source::persist_source: persist_source u5/s82a384a1-c68c-4383-b7a2-8ce47375c516, initial as_of: Some(Antichain { elements: [1666267086554] })
2022-10-20T11:57:59.776315Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086554, 1)
2022-10-20T11:57:59.776341Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267086554, 1)
2022-10-20T11:57:59.776614Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086554, -1)
2022-10-20T11:57:59.776632Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(3), Int32(3), Int32(3), Int32(3)]}), 1666267086554, 1)
2022-10-20T11:57:59.880644Z  INFO mz_storage::source::persist_source: persist_source u5/s82a384a1-c68c-4383-b7a2-8ce47375c516, initial as_of: Some(Antichain { elements: [1666267086554] })
2022-10-20T11:57:59.886930Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086554, 1)
2022-10-20T11:57:59.886953Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267086554, 1)
2022-10-20T11:57:59.887210Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086554, -1)
2022-10-20T11:57:59.887224Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(3), Int32(3), Int32(3), Int32(3)]}), 1666267086554, 1)
2022-10-20T11:58:00.050075Z  INFO mz_storage::source::persist_source: persist_source u5/s82a384a1-c68c-4383-b7a2-8ce47375c516, initial as_of: Some(Antichain { elements: [1666267086554] })
2022-10-20T11:58:00.063657Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086554, 1)
2022-10-20T11:58:00.063688Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267086554, 1)
2022-10-20T11:58:00.065723Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086554, -1)
2022-10-20T11:58:00.065742Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(3), Int32(3), Int32(3), Int32(3)]}), 1666267086554, 1)
2022-10-20T11:58:00.304347Z  INFO mz_storage::source::persist_source: persist_source u5/s82a384a1-c68c-4383-b7a2-8ce47375c516, initial as_of: Some(Antichain { elements: [1666267086557] })
2022-10-20T11:58:00.375712Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086557] } not yet available for upper Antichain { elements: [1666267086555] } retrying in 64ms
2022-10-20T11:58:00.436289Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086557] } not yet available for upper Antichain { elements: [1666267086555] } retrying in 128ms
2022-10-20T11:58:00.578364Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086557] } not yet available for upper Antichain { elements: [1666267086555] } retrying in 256ms
2022-10-20T11:58:00.815079Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086557] } not yet available for upper Antichain { elements: [1666267086555] } retrying in 512ms
2022-10-20T11:58:01.369696Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086557] } not yet available for upper Antichain { elements: [1666267086555] } retrying in 1.024s
2022-10-20T11:58:02.373026Z  INFO mz_persist_client::internal::machine: snapshot s82a384a1-c68c-4383-b7a2-8ce47375c516 as of Antichain { elements: [1666267086557] } not yet available for upper Antichain { elements: [1666267086555] } retrying in 2.048s
2022-10-20T11:58:03.319887Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267071555, -1)
2022-10-20T11:58:03.319915Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(3), Int32(3), Int32(3), Int32(3)]}), 1666267071603, 1)
2022-10-20T11:58:03.319940Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: persist updates: [(Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267071555, -1), (Ok(Row{[Int32(3), Int32(3), Int32(3), Int32(3)]}), 1666267071603, 1)], in-flight-batches: {}, desired_frontier: Antichain { elements: [1666267086555] }, batch_descriptions_frontier: Antichain { elements: [1666267071556] }, persist_frontier: Antichain { elements: [1666267071555] }
2022-10-20T11:58:03.717217Z  INFO mz_compute::sink::persist_sink: persist_sink u5/s82a384a1-c68c-4383-b7a2-8ce47375c516: append result for batch (Antichain { elements: [1666267086555] } -> Antichain { elements: [1666267086558] }): Ok(())
2022-10-20T11:58:04.585215Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086557, 1)
2022-10-20T11:58:04.585245Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(2), Int32(2), Int32(2), Int32(2)]}), 1666267086557, 1)
2022-10-20T11:58:04.586044Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(1), Int32(1), Int32(1), Int32(1)]}), 1666267086557, -1)
2022-10-20T11:58:04.586062Z  INFO mz_storage::source::persist_source: persist_source u5, emitting update: (Ok(Row{[Int32(3), Int32(3), Int32(3), Int32(3)]}), 1666267086557, 1)

```

**Comment by philip-stoev** [comment]
Keeps happening, this one is from today https://buildkite.com/materialize/nightlies/builds/1277#018412e6-b124-48b9-bf5c-f12efc6dd864

**Comment by teskje** [comment]
@philip-stoev Can you still reproduce this today? I have had your repro running for two hours now and so far it's chugging along happily.

**Comment by philip-stoev** [comment]
Yes, this remains reproducible it happens very frequently in Nightly -- e.g. it happened two times in a row today https://buildkite.com/materialize/nightlies/builds/1273#01840f21-dc63-470b-b7fc-70824bed67aa

**Comment by philip-stoev** [comment]
@teskje I suggest you try a scratch machine of type `ci`

**Comment by frankmcsherry** [comment]
Fwiw, I've been running https://github.com/MaterializeInc/materialize/pull/15751 for over an hour now and have been unable to get a reproduction. I'd like to get a better explanation than that, but there are a bunch that might pan out (e.g. incorrect reconciliation allows the use of indexes that have been compacted past a dataflow's `as_of`).

**Comment by teskje** [comment]
If #15751 doesn't break anything in CI, how about we merge it? It looks strictly more correct than what we have now. I have [some doubts](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1666971766490049?thread_ts=1666903553.761499&cid=C02PPB50ZHS) about the claim that we do anything to advance updates to the new `as_of`, but I can also not think of anything that would go immediately wrong if we don't.

We should still continue to debug the issue here though, to get that better explanation. It's possible that #15751 only hides the bug and doesn't fix it.

**Comment by teskje** [comment]
Explanation for why the bug likely happens: [Slack thread](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1667912206665429?thread_ts=1667810329.507299&cid=C02PPB50ZHS).

TLDR: When environmentd restarts, it can select `as_of`s for index dataflows that are before the times the indexes have previously been allowed to compacted to. We had a bug in compute reconciliation that would allow the reuse of a dataflow with an `as_of` beyond the one requested by the compute controller. So upon restart the compute controller would instruct the creation of an index at a certain `as_of`, but the replica would install (or rather: keep running) a dataflow whose outputs had that `as_of` time already compacted. Then we would install a join dataflow on top of two of these invalid indexes. The delta join would read confusing/inconsistent times from its input arrangements, which would cause it to swallow some rows that it should have emitted instead.

#15751 fixes the bug in reconciliation. So instead of keeping the old index dataflows around, the replica drops them and creates new ones that don't have compacted outputs. The delta join is instantiated over valid inputs and functions as expected.

**Comment by teskje** [comment]
Fixed by https://github.com/MaterializeInc/materialize/pull/15751.
